pyfar/dsp/regularization.py

```diff
@@ -0,0 +1,408 @@
+
+        Examples
+        --------
+        Invert a headphone transfer function (HpTF), regularize the inversion at
+        high frequencies, and use a band-pass as target function. Note that the
+        equalized HpTF, which is obtained from a convolution of the HpTF with its
+        inverse filter, approaches the target function in both, time and frequency.
+
+        .. plot::
+
+            >>> import pyfar as pf
+            >>> import numpy as np
+            >>>
+            >>> # Headphone transfer function to be inverted
+            >>> hptf = pf.signals.files.headphone_impulse_responses()[0, 0].flatten()
+            >>> # Regularize inversion at high frequencies
+            >>> regularization = pf.dsp.filter.low_shelf(
+            ...     pf.signals.impulse(hptf.n_samples), 4e3, -20, 2, 'II')
+            >>> # Use band-pass as target function
+            >>> target = pf.dsp.filter.butterworth(
+            ...     pf.signals.impulse(hptf.n_samples), 4, [100, 16e3], 'bandpass')
+            >>> target = pf.dsp.time_shift(target, 125, 'cyclic')
+            >>>
+            >>> # Regulated inversion
+            >>> Inversion = pf.dsp.RegularizedSpectrumInversion.from_magnitude_spectrum(
+            ...     hptf, regularization, beta=.1, target=target)
+            >>> inverted = Inversion.invert
+            >>>
+            >>> # Plot the result
+            >>> ax = pf.plot.time_freq(hptf, label='HpTF')
+            >>> pf.plot.time_freq(inverted, label='HpTF inverted')
+            >>> pf.plot.time_freq(hptf * inverted, label='Equalized HpTF')
+            >>> pf.plot.time_freq(target, linestyle='--', label='Target')
+            >>> pf.plot.freq(regularization, ax=ax[1], label='Regularization')
+            >>> ax[0].set_xlim(0 , .005)
+            >>> ax[1].set_ylim(-40 , 15)
+            >>> ax[1].legend(loc='lower center', ncols=3)
+        """# noqa: E501
+        if not isinstance(regularization, pf.Signal):
+            raise ValueError(
+                "Regularization must be a pyfar.Signal object.")
+        if signal.n_samples != regularization.n_samples:
+            raise ValueError(
+                "The number of samples in the signal and the regularization "
+                "function differs but must be equal.")
+        if signal.sampling_rate != regularization.sampling_rate:
+            raise ValueError(
+                "The sampling rate of the signal and the regularization "
+                "function differs but must be equal.")
+
+        instance = cls.__new__(cls)
+        instance._regularization_signal = regularization
+        instance._regularization_type = "magnitude spectrum"
+        instance.signal = signal
+        instance.beta = beta
+        instance.target = target
+
+        return instance
+
+    @property
+    def beta(self):
+        r"""
+        Amount of regularization :math:`\beta`.
+        """
+        return self._beta
+
+    @beta.setter
+    def beta(self, beta):
```

Comment:
here it smells like a similar behviour as the Filter.state. Do we really want to have a setter were we allow numbers as well as strings defining a certain beahviour? maybe it would be more elegant to allow a number to set and to define an other method to set beta for energy', 'mean', 'max'.